// @aws-amplify/client-config/lib/client-config-contributor/graphql_client_config_contributor.js
import { graphqlOutputKey, } from '@aws-amplify/backend-output-schemas';
/**
 * Translator for the Graphql API portion of ClientConfig
 */
export class GraphqlClientConfigContributor {
    modelIntrospectionSchemaAdapter;
    /**
     * Constructor
     * @param modelIntrospectionSchemaAdapter the adapter to provide the model introspection schema from s3 uri
     */
    constructor(modelIntrospectionSchemaAdapter) {
        this.modelIntrospectionSchemaAdapter = modelIntrospectionSchemaAdapter;
    }
    /**
     * Given some BackendOutput, contribute the Graphql API portion of the client config
     */
    contribute = async ({ [graphqlOutputKey]: graphqlOutput, }) => {
        if (graphqlOutput === undefined) {
            return {};
        }
        const config = {
            aws_appsync_graphqlEndpoint: graphqlOutput.payload.awsAppsyncApiEndpoint,
            aws_appsync_region: graphqlOutput.payload.awsAppsyncRegion,
            aws_appsync_apiKey: graphqlOutput.payload.awsAppsyncApiKey,
            aws_appsync_authenticationType: graphqlOutput.payload.awsAppsyncAuthenticationType,
            aws_appsync_additionalAuthenticationTypes: graphqlOutput.payload.awsAppsyncAdditionalAuthenticationTypes,
            aws_appsync_conflictResolutionMode: graphqlOutput.payload.awsAppsyncConflictResolutionMode,
        };
        const modelIntrospection = await this.modelIntrospectionSchemaAdapter.getModelIntrospectionSchemaFromS3Uri(graphqlOutput.payload.amplifyApiModelSchemaS3Uri);
        if (modelIntrospection) {
            config.modelIntrospection = modelIntrospection;
        }
        return config;
    };
}
