// @aws-amplify/graphql-transformer-interfaces/lib/transformer-context/transformer-datasource-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLLambdaModelProvisionStrategy = exports.DynamoDBProvisionStrategy = exports.AppSyncDataSourceType = void 0;
var AppSyncDataSourceType;
(function (AppSyncDataSourceType) {
    AppSyncDataSourceType["AMAZON_DYNAMODB"] = "AMAZON_DYNAMODB";
    AppSyncDataSourceType["AMAZON_ELASTICSEARCH"] = "AMAZON_ELASTICSEARCH";
    AppSyncDataSourceType["AWS_LAMBDA"] = "AWS_LAMBDA";
    AppSyncDataSourceType["RELATIONAL_DATABASE"] = "RELATIONAL_DATABASE";
    AppSyncDataSourceType["HTTP"] = "HTTP";
    AppSyncDataSourceType["NONE"] = "NONE";
})(AppSyncDataSourceType = exports.AppSyncDataSourceType || (exports.AppSyncDataSourceType = {}));
var DynamoDBProvisionStrategy;
(function (DynamoDBProvisionStrategy) {
    DynamoDBProvisionStrategy["DEFAULT"] = "DEFAULT";
    DynamoDBProvisionStrategy["AMPLIFY_TABLE"] = "AMPLIFY_TABLE";
})(DynamoDBProvisionStrategy = exports.DynamoDBProvisionStrategy || (exports.DynamoDBProvisionStrategy = {}));
var SQLLambdaModelProvisionStrategy;
(function (SQLLambdaModelProvisionStrategy) {
    SQLLambdaModelProvisionStrategy["DEFAULT"] = "DEFAULT";
})(SQLLambdaModelProvisionStrategy = exports.SQLLambdaModelProvisionStrategy || (exports.SQLLambdaModelProvisionStrategy = {}));
