// @aws-sdk/client-amplifyuibuilder/dist-cjs/AmplifyUIBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyUIBuilder = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const AmplifyUIBuilderClient_1 = require("./AmplifyUIBuilderClient");
const CreateComponentCommand_1 = require("./commands/CreateComponentCommand");
const CreateFormCommand_1 = require("./commands/CreateFormCommand");
const CreateThemeCommand_1 = require("./commands/CreateThemeCommand");
const DeleteComponentCommand_1 = require("./commands/DeleteComponentCommand");
const DeleteFormCommand_1 = require("./commands/DeleteFormCommand");
const DeleteThemeCommand_1 = require("./commands/DeleteThemeCommand");
const ExchangeCodeForTokenCommand_1 = require("./commands/ExchangeCodeForTokenCommand");
const ExportComponentsCommand_1 = require("./commands/ExportComponentsCommand");
const ExportFormsCommand_1 = require("./commands/ExportFormsCommand");
const ExportThemesCommand_1 = require("./commands/ExportThemesCommand");
const GetCodegenJobCommand_1 = require("./commands/GetCodegenJobCommand");
const GetComponentCommand_1 = require("./commands/GetComponentCommand");
const GetFormCommand_1 = require("./commands/GetFormCommand");
const GetMetadataCommand_1 = require("./commands/GetMetadataCommand");
const GetThemeCommand_1 = require("./commands/GetThemeCommand");
const ListCodegenJobsCommand_1 = require("./commands/ListCodegenJobsCommand");
const ListComponentsCommand_1 = require("./commands/ListComponentsCommand");
const ListFormsCommand_1 = require("./commands/ListFormsCommand");
const ListThemesCommand_1 = require("./commands/ListThemesCommand");
const PutMetadataFlagCommand_1 = require("./commands/PutMetadataFlagCommand");
const RefreshTokenCommand_1 = require("./commands/RefreshTokenCommand");
const StartCodegenJobCommand_1 = require("./commands/StartCodegenJobCommand");
const UpdateComponentCommand_1 = require("./commands/UpdateComponentCommand");
const UpdateFormCommand_1 = require("./commands/UpdateFormCommand");
const UpdateThemeCommand_1 = require("./commands/UpdateThemeCommand");
const commands = {
    CreateComponentCommand: CreateComponentCommand_1.CreateComponentCommand,
    CreateFormCommand: CreateFormCommand_1.CreateFormCommand,
    CreateThemeCommand: CreateThemeCommand_1.CreateThemeCommand,
    DeleteComponentCommand: DeleteComponentCommand_1.DeleteComponentCommand,
    DeleteFormCommand: DeleteFormCommand_1.DeleteFormCommand,
    DeleteThemeCommand: DeleteThemeCommand_1.DeleteThemeCommand,
    ExchangeCodeForTokenCommand: ExchangeCodeForTokenCommand_1.ExchangeCodeForTokenCommand,
    ExportComponentsCommand: ExportComponentsCommand_1.ExportComponentsCommand,
    ExportFormsCommand: ExportFormsCommand_1.ExportFormsCommand,
    ExportThemesCommand: ExportThemesCommand_1.ExportThemesCommand,
    GetCodegenJobCommand: GetCodegenJobCommand_1.GetCodegenJobCommand,
    GetComponentCommand: GetComponentCommand_1.GetComponentCommand,
    GetFormCommand: GetFormCommand_1.GetFormCommand,
    GetMetadataCommand: GetMetadataCommand_1.GetMetadataCommand,
    GetThemeCommand: GetThemeCommand_1.GetThemeCommand,
    ListCodegenJobsCommand: ListCodegenJobsCommand_1.ListCodegenJobsCommand,
    ListComponentsCommand: ListComponentsCommand_1.ListComponentsCommand,
    ListFormsCommand: ListFormsCommand_1.ListFormsCommand,
    ListThemesCommand: ListThemesCommand_1.ListThemesCommand,
    PutMetadataFlagCommand: PutMetadataFlagCommand_1.PutMetadataFlagCommand,
    RefreshTokenCommand: RefreshTokenCommand_1.RefreshTokenCommand,
    StartCodegenJobCommand: StartCodegenJobCommand_1.StartCodegenJobCommand,
    UpdateComponentCommand: UpdateComponentCommand_1.UpdateComponentCommand,
    UpdateFormCommand: UpdateFormCommand_1.UpdateFormCommand,
    UpdateThemeCommand: UpdateThemeCommand_1.UpdateThemeCommand,
};
class AmplifyUIBuilder extends AmplifyUIBuilderClient_1.AmplifyUIBuilderClient {
}
exports.AmplifyUIBuilder = AmplifyUIBuilder;
(0, smithy_client_1.createAggregatedClient)(commands, AmplifyUIBuilder);
