// @aws-amplify/backend-cli/lib/commands/sandbox/sandbox-delete/sandbox_delete_command.js
import { AmplifyPrompter } from '@aws-amplify/cli-core';
import { handleCommandFailure } from '../../../command_failure_handler.js';
/**
 * Command that deletes the sandbox environment.
 */
export class SandboxDeleteCommand {
    sandboxFactory;
    /**
     * @inheritDoc
     */
    command;
    /**
     * @inheritDoc
     */
    describe;
    /**
     * Deletes sandbox environment.
     */
    constructor(sandboxFactory) {
        this.sandboxFactory = sandboxFactory;
        this.command = 'delete';
        this.describe = 'Deletes sandbox environment';
    }
    /**
     * @inheritDoc
     */
    handler = async (args) => {
        let isConfirmed = args.yes;
        if (!isConfirmed) {
            isConfirmed = await AmplifyPrompter.yesOrNo({
                message: "Are you sure you want to delete all the resources in your sandbox environment (This can't be undone)?",
            });
        }
        if (isConfirmed) {
            await (await this.sandboxFactory.getInstance()).delete({ name: args.name });
        }
    };
    /**
     * @inheritDoc
     */
    builder = (yargs) => {
        return yargs
            .option('yes', {
            describe: 'Do not ask for confirmation before deleting the sandbox environment',
            type: 'boolean',
            array: false,
            alias: 'y',
        })
            .option('name', {
            describe: 'An optional name to distinguish between different sandbox environments. Default is the name in your package.json',
            type: 'string',
            array: false,
        })
            .check((argv) => {
            if (argv.name) {
                const projectNameRegex = /^[a-zA-Z0-9-]{1,15}$/;
                if (!argv.name.match(projectNameRegex)) {
                    throw new Error(`--name should match [a-zA-Z0-9-] and less than 15 characters.`);
                }
            }
            return true;
        })
            .fail((msg, err) => {
            handleCommandFailure(msg, err, yargs);
            yargs.exit(1, err);
        });
    };
}
