// @aws-amplify/backend-auth/lib/factory.js
import { AmplifyAuth, } from '@aws-amplify/auth-construct-alpha';
import * as path from 'path';
import { translateToAuthConstructLoginWith } from './translate_auth_props.js';
/**
 * Singleton factory for AmplifyAuth that can be used in Amplify project files
 */
class AmplifyAuthFactory {
    props;
    importStack;
    provides = 'AuthResources';
    generator;
    /**
     * Set the properties that will be used to initialize AmplifyAuth
     */
    constructor(props, importStack = new Error().stack) {
        this.props = props;
        this.importStack = importStack;
    }
    /**
     * Get a singleton instance of AmplifyAuth
     */
    getInstance = (getInstanceProps) => {
        const { constructContainer, importPathVerifier } = getInstanceProps;
        importPathVerifier?.verify(this.importStack, path.join('amplify', 'auth', 'resource'), 'Amplify Auth must be defined in amplify/auth/resource.ts');
        if (!this.generator) {
            this.generator = new AmplifyAuthGenerator(this.props, getInstanceProps);
        }
        return constructContainer.getOrCompute(this.generator);
    };
}
class AmplifyAuthGenerator {
    props;
    getInstanceProps;
    resourceGroupName = 'auth';
    defaultName = 'amplifyAuth';
    constructor(props, getInstanceProps) {
        this.props = props;
        this.getInstanceProps = getInstanceProps;
    }
    generateContainerEntry = (scope, backendSecretResolver) => {
        const authProps = {
            ...this.props,
            loginWith: translateToAuthConstructLoginWith(this.props.loginWith, backendSecretResolver),
            outputStorageStrategy: this.getInstanceProps.outputStorageStrategy,
        };
        const authConstruct = new AmplifyAuth(scope, this.defaultName, authProps);
        Object.entries(this.props.triggers || {}).forEach(([triggerEvent, handlerFactory]) => {
            authConstruct.addTrigger(triggerEvent, // this type assertion is necessary before .forEach types keys as just "string"
            handlerFactory.getInstance(this.getInstanceProps));
        });
        return authConstruct;
    };
}
/**
 * Provide the settings that will be used for authentication.
 */
export const defineAuth = (props) => 
// Creates a factory that implements ConstructFactory<AmplifyAuth & ResourceProvider<AuthResources>>
new AmplifyAuthFactory(props, new Error().stack);
