// @aws-amplify/backend-cli/lib/form-generation/form_generation_handler.js
import { createLocalGraphqlFormGenerator } from '@aws-amplify/form-generator';
import { createGraphqlDocumentGenerator } from '@aws-amplify/model-generator';
/**
 * Creates a handler for FormGeneration
 */
export class FormGenerationHandler {
    formGenParams;
    /**
     * Instantiates the handler
     */
    constructor(formGenParams) {
        this.formGenParams = formGenParams;
    }
    generate = async (params) => {
        const { backendIdentifier, modelsOutDir, uiOutDir, apiUrl, modelsFilter } = params;
        const { credentialProvider } = this.formGenParams;
        const graphqlClientGenerator = createGraphqlDocumentGenerator({
            backendIdentifier,
            credentialProvider,
        });
        const modelsResult = await graphqlClientGenerator.generateModels({
            language: 'typescript',
        });
        await modelsResult.writeToDirectory(modelsOutDir);
        const localFormGenerator = createLocalGraphqlFormGenerator({
            introspectionSchemaUrl: apiUrl,
            graphqlModelDirectoryPath: './graphql',
        });
        const result = await localFormGenerator.generateForms({
            models: modelsFilter,
        });
        await result.writeToDirectory(uiOutDir);
    };
}
