// @aws-amplify/data-schema/lib-esm/src/CustomOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.subscription = exports.mutation = exports.query = exports.CustomOperationNames = void 0;
const brandName = 'customOperation';
exports.CustomOperationNames = [
    'Query',
    'Mutation',
    'Subscription',
];
function brandedBuilder(builder) {
    return builder;
}
function _custom(typeName) {
    const data = {
        arguments: {},
        returnType: null,
        functionRef: null,
        authorization: [],
        typeName: typeName,
    };
    const builder = brandedBuilder({
        arguments(args) {
            data.arguments = args;
            return this;
        },
        returns(returnType) {
            data.returnType = returnType;
            return this;
        },
        function(functionRefOrName) {
            data.functionRef = functionRefOrName;
            return this;
        },
        authorization(rules) {
            data.authorization = rules;
            return this;
        },
    });
    return { ...builder, data };
}
function query() {
    return _custom('Query');
}
exports.query = query;
function mutation() {
    return _custom('Mutation');
}
exports.mutation = mutation;
function subscription() {
    return _custom('Subscription');
}
exports.subscription = subscription;
