// @aws-amplify/graphql-api-construct/lib/internal/asset-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetManager = void 0;
const fs = require("fs");
const path = require("path");
const cdk = require("aws-cdk-lib");
const TEMP_PREFIX = 'transformer-assets';
const FUNCTION_PREFIX = 'functions';
const RESOLVER_PREFIX = 'resolvers';
/**
 * The asset manager bridges the gap between creation of file assets in the transformer (which provide a name+contents tuple)
 * with the path method which is used in CDK.
 */
class AssetManager {
    constructor() {
        this.tempAssetDir = cdk.FileSystem.mkdtemp(TEMP_PREFIX);
        this.resolverAssets = {};
    }
    addAsset(fileName, contents) {
        this.trackResolverAsset(fileName, contents);
        const writableContents = this.isContentsAReference(fileName) ? this.dereferenceContents(contents) : contents;
        const filePath = path.join(this.tempAssetDir, fileName);
        const fileDirName = path.dirname(filePath);
        if (!fs.existsSync(fileDirName)) {
            fs.mkdirSync(fileDirName, { recursive: true });
        }
        fs.writeFileSync(filePath, writableContents);
        return filePath;
    }
    isContentsAReference(fileName) {
        return fileName.startsWith(FUNCTION_PREFIX);
    }
    dereferenceContents(contents) {
        return fs.readFileSync(contents);
    }
    trackResolverAsset(fileName, contents) {
        if (fileName.startsWith(RESOLVER_PREFIX)) {
            const resolverFileName = fileName.split('/')[1];
            this.resolverAssets[resolverFileName] = contents;
        }
    }
}
exports.AssetManager = AssetManager;
