// @aws-sdk/client-amplify/dist-cjs/Amplify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Amplify = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const AmplifyClient_1 = require("./AmplifyClient");
const CreateAppCommand_1 = require("./commands/CreateAppCommand");
const CreateBackendEnvironmentCommand_1 = require("./commands/CreateBackendEnvironmentCommand");
const CreateBranchCommand_1 = require("./commands/CreateBranchCommand");
const CreateDeploymentCommand_1 = require("./commands/CreateDeploymentCommand");
const CreateDomainAssociationCommand_1 = require("./commands/CreateDomainAssociationCommand");
const CreateWebhookCommand_1 = require("./commands/CreateWebhookCommand");
const DeleteAppCommand_1 = require("./commands/DeleteAppCommand");
const DeleteBackendEnvironmentCommand_1 = require("./commands/DeleteBackendEnvironmentCommand");
const DeleteBranchCommand_1 = require("./commands/DeleteBranchCommand");
const DeleteDomainAssociationCommand_1 = require("./commands/DeleteDomainAssociationCommand");
const DeleteJobCommand_1 = require("./commands/DeleteJobCommand");
const DeleteWebhookCommand_1 = require("./commands/DeleteWebhookCommand");
const GenerateAccessLogsCommand_1 = require("./commands/GenerateAccessLogsCommand");
const GetAppCommand_1 = require("./commands/GetAppCommand");
const GetArtifactUrlCommand_1 = require("./commands/GetArtifactUrlCommand");
const GetBackendEnvironmentCommand_1 = require("./commands/GetBackendEnvironmentCommand");
const GetBranchCommand_1 = require("./commands/GetBranchCommand");
const GetDomainAssociationCommand_1 = require("./commands/GetDomainAssociationCommand");
const GetJobCommand_1 = require("./commands/GetJobCommand");
const GetWebhookCommand_1 = require("./commands/GetWebhookCommand");
const ListAppsCommand_1 = require("./commands/ListAppsCommand");
const ListArtifactsCommand_1 = require("./commands/ListArtifactsCommand");
const ListBackendEnvironmentsCommand_1 = require("./commands/ListBackendEnvironmentsCommand");
const ListBranchesCommand_1 = require("./commands/ListBranchesCommand");
const ListDomainAssociationsCommand_1 = require("./commands/ListDomainAssociationsCommand");
const ListJobsCommand_1 = require("./commands/ListJobsCommand");
const ListTagsForResourceCommand_1 = require("./commands/ListTagsForResourceCommand");
const ListWebhooksCommand_1 = require("./commands/ListWebhooksCommand");
const StartDeploymentCommand_1 = require("./commands/StartDeploymentCommand");
const StartJobCommand_1 = require("./commands/StartJobCommand");
const StopJobCommand_1 = require("./commands/StopJobCommand");
const TagResourceCommand_1 = require("./commands/TagResourceCommand");
const UntagResourceCommand_1 = require("./commands/UntagResourceCommand");
const UpdateAppCommand_1 = require("./commands/UpdateAppCommand");
const UpdateBranchCommand_1 = require("./commands/UpdateBranchCommand");
const UpdateDomainAssociationCommand_1 = require("./commands/UpdateDomainAssociationCommand");
const UpdateWebhookCommand_1 = require("./commands/UpdateWebhookCommand");
const commands = {
    CreateAppCommand: CreateAppCommand_1.CreateAppCommand,
    CreateBackendEnvironmentCommand: CreateBackendEnvironmentCommand_1.CreateBackendEnvironmentCommand,
    CreateBranchCommand: CreateBranchCommand_1.CreateBranchCommand,
    CreateDeploymentCommand: CreateDeploymentCommand_1.CreateDeploymentCommand,
    CreateDomainAssociationCommand: CreateDomainAssociationCommand_1.CreateDomainAssociationCommand,
    CreateWebhookCommand: CreateWebhookCommand_1.CreateWebhookCommand,
    DeleteAppCommand: DeleteAppCommand_1.DeleteAppCommand,
    DeleteBackendEnvironmentCommand: DeleteBackendEnvironmentCommand_1.DeleteBackendEnvironmentCommand,
    DeleteBranchCommand: DeleteBranchCommand_1.DeleteBranchCommand,
    DeleteDomainAssociationCommand: DeleteDomainAssociationCommand_1.DeleteDomainAssociationCommand,
    DeleteJobCommand: DeleteJobCommand_1.DeleteJobCommand,
    DeleteWebhookCommand: DeleteWebhookCommand_1.DeleteWebhookCommand,
    GenerateAccessLogsCommand: GenerateAccessLogsCommand_1.GenerateAccessLogsCommand,
    GetAppCommand: GetAppCommand_1.GetAppCommand,
    GetArtifactUrlCommand: GetArtifactUrlCommand_1.GetArtifactUrlCommand,
    GetBackendEnvironmentCommand: GetBackendEnvironmentCommand_1.GetBackendEnvironmentCommand,
    GetBranchCommand: GetBranchCommand_1.GetBranchCommand,
    GetDomainAssociationCommand: GetDomainAssociationCommand_1.GetDomainAssociationCommand,
    GetJobCommand: GetJobCommand_1.GetJobCommand,
    GetWebhookCommand: GetWebhookCommand_1.GetWebhookCommand,
    ListAppsCommand: ListAppsCommand_1.ListAppsCommand,
    ListArtifactsCommand: ListArtifactsCommand_1.ListArtifactsCommand,
    ListBackendEnvironmentsCommand: ListBackendEnvironmentsCommand_1.ListBackendEnvironmentsCommand,
    ListBranchesCommand: ListBranchesCommand_1.ListBranchesCommand,
    ListDomainAssociationsCommand: ListDomainAssociationsCommand_1.ListDomainAssociationsCommand,
    ListJobsCommand: ListJobsCommand_1.ListJobsCommand,
    ListTagsForResourceCommand: ListTagsForResourceCommand_1.ListTagsForResourceCommand,
    ListWebhooksCommand: ListWebhooksCommand_1.ListWebhooksCommand,
    StartDeploymentCommand: StartDeploymentCommand_1.StartDeploymentCommand,
    StartJobCommand: StartJobCommand_1.StartJobCommand,
    StopJobCommand: StopJobCommand_1.StopJobCommand,
    TagResourceCommand: TagResourceCommand_1.TagResourceCommand,
    UntagResourceCommand: UntagResourceCommand_1.UntagResourceCommand,
    UpdateAppCommand: UpdateAppCommand_1.UpdateAppCommand,
    UpdateBranchCommand: UpdateBranchCommand_1.UpdateBranchCommand,
    UpdateDomainAssociationCommand: UpdateDomainAssociationCommand_1.UpdateDomainAssociationCommand,
    UpdateWebhookCommand: UpdateWebhookCommand_1.UpdateWebhookCommand,
};
class Amplify extends AmplifyClient_1.AmplifyClient {
}
exports.Amplify = Amplify;
(0, smithy_client_1.createAggregatedClient)(commands, Amplify);
