// @aws-amplify/backend-storage/lib/factory.js
import { AmplifyStorage, } from '@aws-amplify/storage-construct-alpha';
import * as path from 'path';
/**
 * Singleton factory for a Storage bucket that can be used in `resource.ts` files
 */
class AmplifyStorageFactory {
    props;
    importStack;
    generator;
    /**
     * Set the properties that will be used to initialize the bucket
     */
    constructor(props, importStack = new Error().stack) {
        this.props = props;
        this.importStack = importStack;
    }
    /**
     * Get a singleton instance of the Bucket
     */
    getInstance = ({ constructContainer, outputStorageStrategy, importPathVerifier, }) => {
        importPathVerifier?.verify(this.importStack, path.join('amplify', 'storage', 'resource'), 'Amplify Storage must be defined in amplify/storage/resource.ts');
        if (!this.generator) {
            this.generator = new AmplifyStorageGenerator(this.props, outputStorageStrategy);
        }
        return constructContainer.getOrCompute(this.generator);
    };
}
class AmplifyStorageGenerator {
    props;
    outputStorageStrategy;
    resourceGroupName = 'storage';
    defaultName = 'amplifyStorage';
    constructor(props, outputStorageStrategy) {
        this.props = props;
        this.outputStorageStrategy = outputStorageStrategy;
    }
    generateContainerEntry = (scope) => {
        return new AmplifyStorage(scope, this.defaultName, {
            ...this.props,
            outputStorageStrategy: this.outputStorageStrategy,
        });
    };
}
/**
 * Creates a factory that implements ConstructFactory<AmplifyStorage>
 */
export const defineStorage = (props) => new AmplifyStorageFactory(props, new Error().stack);
